src/metadata/index.rs

```diff
@@ -0,0 +1,116 @@
+use elasticsearch::cat::CatIndicesParts;
+use elasticsearch::params::Time;
+use elasticsearch::{Elasticsearch, Error};
+use std::collections::HashMap;
+use tokio::sync::RwLock;
+
+use crate::{exporter_metrics::SUBSYSTEM_REQ_HISTOGRAM, Exporter};
+
+pub(crate) type IndexToMetadata = RwLock<IndexDataMap>;
+
+pub(crate) type IndexDataMap = HashMap<String, IndexMetadata>;
+
+/// Current Index metadata
+#[derive(Debug, Clone)]
+pub struct IndexMetadata {
+    pub last_hearbeat: chrono::DateTime<chrono::Utc>,
+}
+
+impl IndexMetadata {
+    /// Initialize Index metadata
+    pub fn new() -> Self {
+        Self {
+            last_hearbeat: chrono::Utc::now(),
+        }
+    }
+
+    /// Set heartbeat
+    pub fn reset_heartbeat(&mut self) -> &mut Self {
+        self.last_hearbeat = chrono::Utc::now();
+        self
+    }
+}
+
+/// Current Index metadata
+#[derive(Debug, Deserialize, Default)]
+pub(crate) struct CatIndexData {
+    /// index name
+    pub(crate) index: String,
+}
+
+pub(crate) async fn build(client: &Elasticsearch) -> Result<IndexDataMap, Error> {
+    info!("Elasticsearch: fetching indices metadata");
+    let index_data_map = client
+        .cat()
+        .indices(CatIndicesParts::None)
+        .format("json")
+        .h(&["index"])
+        // Return local information, do not retrieve the state from master node (default: false)
+        .local(true)
+        .time(Time::Ms)
+        .send()
+        .await?
+        .json::<Vec<CatIndexData>>()
+        .await?
+        .into_iter()
+        .map(|value| (value.index, IndexMetadata::new()))
+        .collect();
+
+    Ok(index_data_map)
+}
+
+#[inline]
+fn update_map(old: &mut IndexDataMap, new: IndexDataMap) {
+    for (key, new_metadata) in new {
+        let _ = old.entry(key).or_insert(new_metadata).reset_heartbeat();
+    }
+}
+
+#[allow(unused)]
+pub(crate) async fn poll(exporter: Exporter) {
+    let start = tokio::time::Instant::now();
+
+    let mut interval =
+        tokio::time::interval_at(start, exporter.options().exporter_metadata_refresh_interval);
+
+    loop {
+        let _ = interval.tick().await;
```

Comment:
Can you help me to understand how this is being used?

Reply:
Index metadata is not yet used, I will make another pull request which will implement usage.
poll function has a loop that ticks every 180seconds (by default), this functionality might be outdated as tokio improved a lot, the loop block part is nasty but at the time of implementation this was the way.